tirocks/src/option/db.rs

```diff
@@ -0,0 +1,846 @@
+        enable_pipelined_commit: bool
+
+        /// If true, allow multi-writers to update mem tables in parallel.
+        /// Only some memtable_factory-s support concurrent writes; currently it
+        /// is implemented only for SkipListFactory.  Concurrent memtable writes
+        /// are not compatible with inplace_update_support or filter_deletes.
+        /// It is strongly recommended to set enable_write_thread_adaptive_yield
+        /// if you are going to use this feature.
+        ///
+        /// Default: true
+        allow_concurrent_memtable_write: bool
+
+        /// If true, threads synchronizing with the write batch group leader will
+        /// wait for up to write_thread_max_yield_usec before blocking on a mutex.
+        /// This can substantially improve throughput for concurrent workloads,
+        /// regardless of whether allow_concurrent_memtable_write is enabled.
+        ///
+        /// Default: true
+        enable_write_thread_adaptive_yield: bool
+
+        /// Recovery mode to control the consistency while replaying WAL
+        /// Default: kPointInTimeRecovery
+        wal_recovery_mode: WalRecoveryMode
+
+        /// If true WAL is not flushed automatically after each write. Instead it
+        /// relies on manual invocation of FlushWAL to write the WAL buffer to its
+        /// file.
+        manual_wal_flush: bool
+
+        /// If true, RocksDB supports flushing multiple column families and committing
+        /// their results atomically to MANIFEST. Note that it is not
+        /// necessary to set atomic_flush to true if WAL is always enabled since WAL
+        /// allows the database to be restored to the last persistent state in WAL.
+        /// This option is useful when there are column families with writes NOT
+        /// protected by WAL.
+        /// For manual flush, application has to specify which column families to
+        /// flush atomically in DB::Flush.
+        /// For auto-triggered flush, RocksDB atomically flushes ALL column families.
+        ///
+        /// Currently, any WAL-enabled writes after atomic flush may be replayed
+        /// independently if the process crashes later and tries to recover.
+        atomic_flush: bool
+    }
+
+    pub(crate) unsafe fn from_ptr<'a>(ptr: *mut rocksdb_DBOptions) -> &'a RawDbOptions {
+        &*(ptr as *mut RawDbOptions)
+    }
+
+    pub(crate) unsafe fn from_ptr_mut<'a>(ptr: *mut rocksdb_DBOptions) -> &'a mut RawDbOptions {
+        &mut *(ptr as *mut RawDbOptions)
+    }
+
+    pub(crate) fn as_ptr(&self) -> *const rocksdb_DBOptions {
+        self as *const _ as _
+    }
+
+    pub(crate) fn as_mut_ptr(&mut self) -> *mut rocksdb_DBOptions {
+        self as *mut _ as _
+    }
+}
+
+impl DbOptions {
+    /// Same as [`RawDbOptions::set_env`] but manage the lifetime of `env`.
+    #[inline]
+    pub fn set_env(&mut self, env: Arc<Env>) -> &mut Self {
+        unsafe {
+            (**self).set_env(&env);
```

Comment:
Say it panics, `env` will be dropped before options?

Reply:
No, it will be destroyed after options. Technically, options have a pointer to env, so options should be dropped first.

Reply:
Why? I thought in this case the local variable (env) will be destroyed first, then self.

Reply:
Oh, it won't panic when setting env. It can only coredump.

Reply:
Make sense.